### PR TITLE
tests,storage: Fix flaky test_rawkv::test_leader_transfer

### LINKED ISSUE #16789: [flay test]  test_rawkv::test_leader_transfer
https://ci.pingcap.net/blue/organizations/jenkins/tikv_ghpr_test/detail/tikv_ghpr_test/43640/tests/

Stacktrace
thread 'cases::test_rawkv::test_leader_transfer' panicked at '', components/test_raftstore/src/util.rs:1422:5
stack backtrace:
   0: rust_begin_unwind
             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/std/src/panicking.rs:575:5
   1: core::panicking::panic_fmt
             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/core/src/panicking.rs:65:14
   2: test_raftstore::util::must_raw_put
   3: failpoints::cases::test_rawkv::TestSuite::must_raw_put
             at /home/jenkins/tikv-src/tests/failpoints/cases/test_rawkv.rs:65:9
   4: failpoints::cases::test_rawkv::test_leader_transfer
             at /home/jenkins/tikv-src/tests/failpoints/cases/test_rawkv.rs:196:9
   5: failpoints::cases::test_rawkv::test_leader_transfer::{{closure}}
             at /home/jenkins/tikv-src/tests/failpoints/cases/test_rawkv.rs:153:27
   6: core::ops::function::FnOnce::call_once
             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/core/src/ops/function.rs:513:5
   7: test_util::runner::run_test_with_hook::{{closure}}::{{closure}}
   8: core::ops::function::FnOnce::call_once{{vtable.shim}}
   9: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at /rustc/96ddd32c4bfb1d78f0cd03eb068b1710a8cebeef/library/alloc/src/boxed.rs:2000:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

### PR DESCRIPTION
**Pull request by pingyu**
### What is changed and how it works?

Issue Number: Close #16789 

What's Changed:

```commit-message
Add RawKvMaxTimestampNotSynced error and set message to errorpb.Error.max_ts_not_synced to provide more information.
Retry on max_ts_not_synced error for must_raw_put.
```

The debug string of "max timestamp not synced" region error will be, e.g, `message: "RawKV write fails due to potentially stale max timestamp, region_id: 1"`.

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
Add RawKvMaxTimestampNotSynced error and set message to errorpb.Error.max_ts_not_synced to provide more information.
Retry on max_ts_not_synced error for must_raw_put.
```

### CODE REVIEW

**Comment by pingyu** [comment]
> /cc @tonyxuqqi

PTAL, thanks~

File: src/storage/txn/mod.rs [hunk 1]
```diff
@@ -144,6 +144,9 @@ pub enum ErrorInner {
     )]
     MaxTimestampNotSynced { region_id: u64, start_ts: TimeStamp },
 
+    #[error("RawKV write fails due to potentially stale max timestamp, region_id: {region_id}")]
+    RawKvMaxTimestampNotSynced { region_id: u64 },
```
Comment (tonyxuqqi) [inline src/storage/txn/mod.rs:148]: Can MaxTimestampNotSynced be used for this purpose as well?

**Review by tonyxuqqi — APPROVED** [review]

File: src/storage/txn/mod.rs (hunk 1 above)
Comment (pingyu) [reply src/storage/txn/mod.rs:148]: The error message "Prewrite for async commit fails due to potentially stale max timestamp" of `MaxTimestampNotSynced` is not appropriate for RawKV, and I am afraid of changing the error message will break some client codes of test cases.

So I prefer to add a new error enum.

**Review by YuJuncen — APPROVED** [review]

**Pull request merged**